Replaced elements with aspect ratio and size in one dimension should respect min-max constraints in opposite dimension.

sammygill opened the pull request:
#### 036e53a548a5852a4a59f49298cb45e8c8e0a2ce
<pre>
Replaced elements with aspect ratio and size in one dimension should respect min-max constraints in opposite dimension.
<a href="https://bugs.webkit.org/show_bug.cgi?id=247507">https://bugs.webkit.org/show_bug.cgi?id=247507</a>
rdar://101979495

Reviewed by Alan Baradlay.

There are certain scenarios where a replaced element may have a
specified aspect ratio, but will not have intrinsic sizes in both
dimensions. One such examples is a SVG with only a specified width
and aspect ratio.

To help facilitate this, the logic in
RenderReplaced::computeAspectRatioInformationForRenderBox has been
separated out since it was doing too much before. It was both computing
the intrinsic size/ratio and constraining the intrinsic size. The logic
for constraining the intrinsic size has been refactored out into a new
method called
RenderReplaced::computeIntrinsicSizesConstrainedByTransferredMinMaxSizes.
This method will call

In these scenarios, we must use the specified aspect ratio to do any
sort of sizing since it is not possible to compute it from the
dimensions of the box. computeAspectRatioInformationForRenderBox to get
the intrinsic size/ratio and then constrain it. This new method will
also constrain the sizes in each dimension by the transferred sizes.
This size may be then potentially constrained even further in
RenderReplaced::computeReplacedLogicalWidthRespectingMinMaxWidth (and
height) by any definite min/max sizes in that dimension. There was an
attempt to consolidate all constraining logic into that method, but
that was causing some unexpected results. That may need to be done in
a future patch.

A few of the CSS aspect ratio methods have also been modified to help
with this logic. Since a lot of the logic is shared, these methods
can determine if a replaced element is being used and use the correct
aspect ratio in that case.

A new member variable to RenderReplaced has been added:
m_intrinsicRatioSize. This is used to keep track of the width and height
sizes associated with the aspect ratio. This is because there were some
precision issues that this patch exposed. Certain tests were starting
to fail because the computed width/height from the aspect ratio was
slightly off. To help with this, we can store the intrinsic ratio sizes
until we actually need to use them to compute the width/height. The
goal here is to hopefully reduce the number of floating point operations
that are needed to compute a size. It may be possible to replace all
instances where the computed intrinsicRatio double is being used, but
that is probably a little too risky for this patch. Instead, this patch
only uses it in the call to resolveHeightForRatio within
RenderReplaced::computeReplacedLogicalHeight. It is probably best
to update all uses of intrinsicRatio to m_intrinsicRatioSize in another
patch. To compute the intrinsic ratio size we needed to keep track of
a new variable inside the computeIntrinsicRatioInformation logic and
then update the member variable once it was computed.

Spec references:
<a href="https://www.w3.org/TR/CSS22/visudet.html#min-max-widths">https://www.w3.org/TR/CSS22/visudet.html#min-max-widths</a>
<a href="https://www.w3.org/TR/CSS22/visudet.html#min-max-heights">https://www.w3.org/TR/CSS22/visudet.html#min-max-heights</a>
<a href="https://w3c.github.io/csswg-drafts/css-sizing-4/#aspect-ratio">https://w3c.github.io/csswg-drafts/css-sizing-4/#aspect-ratio</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/replaced-element-039-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/replaced-element-039.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/replaced-element-040-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/replaced-element-040.html: Added.
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computeChildPreferredLogicalWidths const):
(WebCore::RenderBlock::availableLogicalHeightForPercentageComputation const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::constrainLogicalMinMaxSizesByAspectRatio const):
(WebCore::RenderBox::constrainLogicalHeightByMinMax const):
(WebCore::inlineSizeFromAspectRatio):
Two extra arguments have been added to provide some context when
determining the box type to use (content box or border box). These
arguments are the aspect-ratio type being used for the box along with
whether or not the box is a replaced element with an intrinsic ratio.
css-sizing-4 specifies which box type to use depending on whether the
aspect-ratio property is auto, &lt;ratio&gt;, or auto &amp;&amp; &lt;ratio&gt;. If the item
is a replaced element, then it should always work with the content box.

(WebCore::RenderBox::computeLogicalHeight const):
(WebCore::RenderBox::availableLogicalHeightUsing const):
(WebCore::RenderBox::computePositionedLogicalHeightUsing const):
(WebCore::RenderBox::computeLogicalWidthFromAspectRatioInternal const):
(WebCore::RenderBox::isRenderReplacedWithIntrinsicRatio const):
(WebCore::RenderBox::resolveAspectRatio const):
A new helper method to determine which aspect ratio to use depending on
the type of the box. Since RenderReplaced objects have slightly
different logic to compute their aspect ratio, this method does the
work of finding on the aspect ratio that needs to be used and performs
the appropriate calls to compute it. If the object is a RenderReplaced
object, then we need the compute its aspect ratio, otherwise we should
be able to get the box&apos;s aspect ratio from its style.

(WebCore::RenderBox::computeMinMaxLogicalWidthFromAspectRatio const):
(WebCore::RenderBox::computeMinMaxLogicalHeightFromAspectRatio const):
horizontalBorderAndPaddingExtent and verticalBorderAndPaddingExtent
were replaced with borderAndPaddingLogicalWidth and
borderAndPaddingLogicalHeight since both blockSizeFromAspectRatio and
inlineSizeFromAspectRatio were expecting the inline and block sizes.
Before, these methods were being passed in the values for the physical
directions and were causing certain tests to fail (e.g.
css/css-flexbox/flex-aspect-ratio-img-column-008.html and
css/css-flexbox/image-as-flexitem-size-007v.html).

* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::blockSizeFromAspectRatio):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::hasIntrinsicAspectRatio const):
This method was made virtual because it was difficult to provide a valid
answer for SVG elements at this high of an abstraction. Since SVG
elements may not have an aspect ratio, we need those elements to
override this method and provide the answer themselves.

* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeAspectRatioInformationForRenderBox const):
(WebCore::RenderReplaced::computeIntrinsicSizesConstrainedByTransferredMinMaxSizes const):
(WebCore::RenderReplaced::computeIntrinsicRatioInformation const):
(WebCore::RenderReplaced::computeReplacedLogicalWidth const):
(WebCore::RenderReplaced::computeReplacedLogicalHeight const):
* Source/WebCore/rendering/RenderReplaced.h:
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::hasIntrinsicAspectRatio const):
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::hasIntrinsicAspectRatio const):
* Source/WebCore/rendering/svg/RenderSVGRoot.h:

Canonical link: <a href="https://commits.webkit.org/257434@main">https://commits.webkit.org/257434@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42e8fd285790dec7926e86070efbfe5935b28476

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98999 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8205 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32155 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108406 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168658 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102961 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8758 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91523 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/106366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104719 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/33655 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2107 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2011 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5119 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6974 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/42551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3425 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cathiechen commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -739,9 +740,9 @@ override;
+    static LayoutUnit blockSizeFromAspectRatio(LayoutUnit borderPaddingInlineSum, LayoutUnit borderPaddingBlockSum, double aspectRatio, BoxSizing boxSizing, LayoutUnit inlineSize, AspectRatioType aspectRatioType, bool isReplacedElementWithIntrinsicRatio)
     {
-        if (boxSizing == BoxSizing::BorderBox)
+        if (boxSizing == BoxSizing::BorderBox && aspectRatioType == AspectRatioType::Ratio && !isReplacedElementWithIntrinsicRatio)
```
Should we consider `auto && [<ratio>]` here?

sammygill replied:
I believe it should be ok here to leave it out. If it auto && ratio, then this check will fail and it will use the ContentBox dimensions. 

However, now that I'm looking at the spec again, maybe this should just be `boxSizing == BoxSizing::BorderBox && aspectRatioType == AspectRatioType::Ratio`? Since the only time we need the BorderBox size is when aspect-ratio is `<ratio>` and the box-sizing property is set to border-box.

Thoughts?

https://w3c.github.io/csswg-drafts/css-sizing-4/#aspect-ratio

cathiechen replied:
Yes, I agree with you! Should be `boxSizing == BoxSizing::BorderBox && aspectRatioType == AspectRatioType::Ratio`.

sammygill replied:
So it looks like the reason I had this additional check is because css-sizing/aspect-ratio/replaced-element-014.html seems to assert that replaced elements should *always* use the content box sizing. Maybe we can change this argument to take a bool and pass in `isRenderReplaced()`?

I think that would allow us to get rid of `isReplacedElementWithIntrinsicRatio` and only add in the necessary logic in the couple places where it is needed.

sammygill replied:
I couldn't completely get rid of the method, but I renamed is to `isRenderReplacedWithAspectRatio`. The method is not used as much as it was before, but it still needs to actually compute the intrinsic ratio. I tried to add the intrinsic ratio as a member field and update it, but it wasn't completely straightforward. It will take a little bit of work to determine when we should actually update the intrinsic ratio and then update it accordingly.

I think that work should probably be done outside the scope of this patch and hopefully we could land this without that logic. If you disagree, though, let me know and I can perhaps see what can be done!

cathiechen commented on Source/WebCore/rendering/svg/LegacyRenderSVGRoot.cpp:
```diff
@@ -69,6 +69,14 @@ SVGSVGElement& LegacyRenderSVGRoot::svgSVGElement() const
+{
+    FloatSize intrinsicSize;
+    double intrinsicRatio = 0;
+    computeIntrinsicRatioInformation(intrinsicSize, intrinsicRatio);
```
Hmm, this looks like a hacking way to me...

sammygill replied:
Since I moved the logic of computing the transferred min/max sizes and constraining the computed size into `computeReplacedLogicalWidthRespectingMinMaxWidth` (and height), we don't need the second method to constrain anymore. We just need `computeIntrinsicRatioInformation` to get the intrinsic ratio and sizes. All of the constraining will then be done in the other method.

sammygill replied:
This is no longer true since the change was causing too many issues. computeReplacedLogicalWidthRespectingMinMaxWidth is only constraining with definite sizes in the same dimension. renderreplaced::computeIntrinsicSizesConstrainedByTransferredMinMaxSizes is constraining using the transferred min/max sizes.

cathiechen commented on Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
```diff
@@ -79,6 +79,14 @@ void RenderSVGRoot::setViewportContainer(RenderSVGViewportContainer& viewportCon
+{
+    FloatSize intrinsicSize;
+    double intrinsicRatio;
+    computeIntrinsicRatioInformation(intrinsicSize, intrinsicRatio);
```
Ditto.

cathiechen commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5421,35 +5421,47 @@ LayoutUnit RenderBox::computeLogicalWidthFromAspectRatio(RenderFragmentContainer
+bool RenderBox::isReplacedElementWithIntrinsicRatio() const
+{
+    if (auto* replaced = dynamicDowncast<RenderReplaced>(this))
+        return replaced->computeIntrinsicAspectRatio();
```
Hmm, this looks like a hacking way to me... And this function is called by many places. Do we have an efficient way to figure out if a replaced element has aspect ratio or not?

sammygill replied:
That's one of the main issues that I have been trying to figure out. Since not all Replaced Elements have an intrinsic ratio, right now they are being computed on demand. 

I think ideally we would store both the intrinsic and ratio as member fields. Then we could have a `RenderBox::hasAspectRatio` method that would check both `style().hasAspectRatio()` and the replaced element fields when needed.

cathiechen replied:
I wonder can we just use `replaced->intrinsicSize().isEmpty()` here?

sammygill replied:
Unfortunately, I don't think that will work for all cases :( For example, a SVG could have a dimension missing but still have an intrinsic ratio. imported/w3c/web-platform-tests/css/css-flexbox/flex-aspect-ratio-img-row-010.html is an example of a test with such behavior

cathiechen commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -2585,9 +2585,9 @@ void RenderBox::updateLogicalWidth()
+static LayoutUnit inlineSizeFromAspectRatio(LayoutUnit borderPaddingInlineSum, LayoutUnit borderPaddingBlockSum, double aspectRatio, BoxSizing boxSizing, LayoutUnit blockSize, AspectRatioType aspectRatioType, bool isReplacedElementWithIntrinsicRatio)
 {
-    if (boxSizing == BoxSizing::BorderBox)
+    if (boxSizing == BoxSizing::BorderBox && aspectRatioType == AspectRatioType::Ratio && !isReplacedElementWithIntrinsicRatio)
```
Ditto. What if the value of aspect-ratio is `auto && <ratio>` ?

cathiechen commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5421,35 +5421,47 @@ LayoutUnit RenderBox::computeLogicalWidthFromAspectRatio(RenderFragmentContainer
-    ASSERT(style().hasAspectRatio());
+    ASSERT(style().hasAspectRatio() || isReplacedElementWithIntrinsicRatio());
+
+    double aspectRatio = isReplacedElementWithIntrinsicRatio() ? dynamicDowncast<RenderReplaced>(this)->computeIntrinsicAspectRatio() : style().logicalAspectRatio();
```
Looks like `computeIntrinsicAspectRatio()` is called twice if it is replacedElement.

sammygill replied:
This should be fixed now. I added a resolveAspectRatio method to help determine the aspect ratio to use and it should only call `computeIntrinsicAspectRatio` once if it is replaced.

cathiechen commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5421,35 +5421,47 @@ LayoutUnit RenderBox::computeLogicalWidthFromAspectRatio(RenderFragmentContainer
-        if (LayoutUnit inlineMinSize = computeLogicalWidthInFragmentUsing(MinSize, style().logicalMinWidth(), availableWidth(), *containingBlock(), nullptr); inlineMinSize > LayoutUnit())
-            transferredMinSize = blockSizeFromAspectRatio(horizontalBorderAndPaddingExtent(), verticalBorderAndPaddingExtent(), style().logicalAspectRatio(), style().boxSizingForAspectRatio(), inlineMinSize);
+        if (LayoutUnit inlineMinSize = computeLogicalWidthInFragmentUsing(MinSize, style().logicalMinWidth(), containingBlockLogicalWidthForContent(), *containingBlock(), nullptr); inlineMinSize > LayoutUnit())
+            transferredMinSize = blockSizeFromAspectRatio(borderAndPaddingLogicalWidth(), borderAndPaddingLogicalHeight(), aspectRatio, style().boxSizingForAspectRatio(), inlineMinSize, style().aspectRatioType(), isReplacedElementWithIntrinsicRatio());
```
Just curious, why do we use `borderAndPaddingLogicalWidth(), borderAndPaddingLogicalHeight()` instead of `horizontalBorderAndPaddingExtent(), verticalBorderAndPaddingExtent()` here?

sammygill replied:
It looks like inlineSizeFromAspectRatio and blockSizeFromAspectRatio expect these two values to be the border+padding of the inline and block sizes. I think it was previously returning the border+padding in terms of physical directions and was causing tests with different writing modes to fail (e.g. css/css-flexbox/flex-aspect-ratio-img-column-008.html	and css/css-flexbox/image-as-flexitem-size-007v.html)

cathiechen commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -425,25 +425,35 @@ void RenderReplaced::computeAspectRatioInformationForRenderBox(RenderBox* conten
+    constrainedSize = intrinsicSize;
+}
 
+void RenderReplaced::constrainIntrinsicSize(FloatSize& intrinsicSize, double& intrinsicRatio) const
```
Feels like using `constrainedSize` is more consistent with `RenderReplaced::computeAspectRatioInformationForRenderBox` and the callers.

sammygill replied:
Do you still think this should be changed with all of the changes `computeIntrinsicRatio` should only provide the intrinsic size now and not do any constraining.

cathiechen commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -425,25 +425,35 @@ void RenderReplaced::computeAspectRatioInformationForRenderBox(RenderBox* conten
+                maxSize = std::max(0_lu, maxSize - borderAndPadding);
+            };
+
+            auto [minLogicalWidth, maxLogicalWidth] = computeMinMaxLogicalWidthFromAspectRatio();
```
Not sure if this is the right place to call `computeMinMaxLogicalWidthFromAspectRatio` which is calculating the transferred min/max width. I think logically, `computeMinMaxLogicalWidthFromAspectRatio` should be used in `RenderBox::computeReplacedLogicalWidthRespectingMinMaxWidth`, which is handling min/max size for replaced elements.
You may take a look at how non-replaced elements work, see `RenderBox::constrainLogicalMinMaxSizesByAspectRatio`.

sammygill replied:
I moved this logic to `RenderBox::computeReplacedLogicalWidthRespectingMinMaxWidth`. I also modified `RenderReplaced::computeReplacedLogicalWidth` (and height) to no longer call these methods directly. Instead, these method are just providing a computed width/height, and then the result is being constrained by the caller so that it looks something like `computeReplacedLogicalWidthRespectingMinMaxHeight(computeReplacedLogicalWidth())`. There was already some code doing this in other parts of the code, too all of these calls should be consistent now. 

Once thing we could potentially do is just call `computeReplacedLogicalWidth` *within* `computeReplacedLogicalWidthRespectingMinMaxWidth,` so that it looks a little cleaner but let me know what you think.

sammygill replied:
This change was causing a lot of issues so I reverted it back to the original version.

cathiechen reviewed (commented):
Hi @sgill26 ! Thanks for working on this!

sammygill committed:
Replaced elements with aspect ratio and size in one dimension should respect min-max constraints in opposite dimension.
https://bugs.webkit.org/show_bug.cgi?id=247507
rdar://101979495

Reviewed by Alan Baradlay.

There are certain scenarios where a replaced element may have a
specified aspect ratio, but will not have intrinsic sizes in both
dimensions. One such examples is a SVG with only a specified width
and aspect ratio.

To help facilitate this, the logic in
RenderReplaced::computeAspectRatioInformationForRenderBox has been
separated out since it was doing too much before. It was both computing
the intrinsic size/ratio and constraining the intrinsic size. The logic
for constraining the intrinsic size has been refactored out into a new
method called
RenderReplaced::computeIntrinsicSizesConstrainedByTransferredMinMaxSizes.
This method will call

In these scenarios, we must use the specified aspect ratio to do any
sort of sizing since it is not possible to compute it from the
dimensions of the box. computeAspectRatioInformationForRenderBox to get
the intrinsic size/ratio and then constrain it. This new method will
also constrain the sizes in each dimension by the transferred sizes.
This size may be then potentially constrained even further in
RenderReplaced::computeReplacedLogicalWidthRespectingMinMaxWidth (and
height) by any definite min/max sizes in that dimension. There was an
attempt to consolidate all constraining logic into that method, but
that was causing some unexpected results. That may need to be done in
a future patch.

A few of the CSS aspect ratio methods have also been modified to help
with this logic. Since a lot of the logic is shared, these methods
can determine if a replaced element is being used and use the correct
aspect ratio in that case.

A new member variable to RenderReplaced has been added:
m_intrinsicRatioSize. This is used to keep track of the width and height
sizes associated with the aspect ratio. This is because there were some
precision issues that this patch exposed. Certain tests were starting
to fail because the computed width/height from the aspect ratio was
slightly off. To help with this, we can store the intrinsic ratio sizes
until we actually need to use them to compute the width/height. The
goal here is to hopefully reduce the number of floating point operations
that are needed to compute a size. It may be possible to replace all
instances where the computed intrinsicRatio double is being used, but
that is probably a little too risky for this patch. Instead, this patch
only uses it in the call to resolveHeightForRatio within
RenderReplaced::computeReplacedLogicalHeight. It is probably best
to update all uses of intrinsicRatio to m_intrinsicRatioSize in another
patch. To compute the intrinsic ratio size we needed to keep track of
a new variable inside the computeIntrinsicRatioInformation logic and
then update the member variable once it was computed.

Spec references:
https://www.w3.org/TR/CSS22/visudet.html#min-max-widths
https://www.w3.org/TR/CSS22/visudet.html#min-max-heights
https://w3c.github.io/csswg-drafts/css-sizing-4/#aspect-ratio

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/replaced-element-039-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/replaced-element-039.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/replaced-element-040-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/replaced-element-040.html: Added.
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computeChildPreferredLogicalWidths const):
(WebCore::RenderBlock::availableLogicalHeightForPercentageComputation const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::constrainLogicalMinMaxSizesByAspectRatio const):
(WebCore::RenderBox::constrainLogicalHeightByMinMax const):
(WebCore::inlineSizeFromAspectRatio):
Two extra arguments have been added to provide some context when
determining the box type to use (content box or border box). These
arguments are the aspect-ratio type being used for the box along with
whether or not the box is a replaced element with an intrinsic ratio.
css-sizing-4 specifies which box type to use depending on whether the
aspect-ratio property is auto, <ratio>, or auto && <ratio>. If the item
is a replaced element, then it should always work with the content box.

(WebCore::RenderBox::computeLogicalHeight const):
(WebCore::RenderBox::availableLogicalHeightUsing const):
(WebCore::RenderBox::computePositionedLogicalHeightUsing const):
(WebCore::RenderBox::computeLogicalWidthFromAspectRatioInternal const):
(WebCore::RenderBox::isRenderReplacedWithIntrinsicRatio const):
(WebCore::RenderBox::resolveAspectRatio const):
A new helper method to determine which aspect ratio to use depending on
the type of the box. Since RenderReplaced objects have slightly
different logic to compute their aspect ratio, this method does the
work of finding on the aspect ratio that needs to be used and performs
the appropriate calls to compute it. If the object is a RenderReplaced
object, then we need the compute its aspect ratio, otherwise we should
be able to get the box's aspect ratio from its style.

(WebCore::RenderBox::computeMinMaxLogicalWidthFromAspectRatio const):
(WebCore::RenderBox::computeMinMaxLogicalHeightFromAspectRatio const):
horizontalBorderAndPaddingExtent and verticalBorderAndPaddingExtent
were replaced with borderAndPaddingLogicalWidth and
borderAndPaddingLogicalHeight since both blockSizeFromAspectRatio and
inlineSizeFromAspectRatio were expecting the inline and block sizes.
Before, these methods were being passed in the values for the physical
directions and were causing certain tests to fail (e.g.
css/css-flexbox/flex-aspect-ratio-img-column-008.html and
css/css-flexbox/image-as-flexitem-size-007v.html).

* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::blockSizeFromAspectRatio):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::hasIntrinsicAspectRatio const):
This method was made virtual because it was difficult to provide a valid
answer for SVG elements at this high of an abstraction. Since SVG
elements may not have an aspect ratio, we need those elements to
override this method and provide the answer themselves.

* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeAspectRatioInformationForRenderBox const):
(WebCore::RenderReplaced::computeIntrinsicSizesConstrainedByTransferredMinMaxSizes const):
(WebCore::RenderReplaced::computeIntrinsicRatioInformation const):
(WebCore::RenderReplaced::computeReplacedLogicalWidth const):
(WebCore::RenderReplaced::computeReplacedLogicalHeight const):
* Source/WebCore/rendering/RenderReplaced.h:
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::hasIntrinsicAspectRatio const):
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::hasIntrinsicAspectRatio const):
* Source/WebCore/rendering/svg/RenderSVGRoot.h:

Canonical link: https://commits.webkit.org/257434@main
